Spigot-Server-Patches/0730-additions-to-PlayerGameModeChangeEvent.patch

```diff
@@ -0,0 +1,126 @@
++                if (event != null && event.isCancelled()) {
++                    commandcontext.getSource().sendMessage(io.papermc.paper.adventure.PaperAdventure.asVanilla(event.cancelMessage()), false);
+                     continue;
+                 }
+-                // CraftBukkit end
++                // Paper end
+                 a((CommandListenerWrapper) commandcontext.getSource(), entityplayer, enumgamemode);
+                 ++i;
+             }
+diff --git a/src/main/java/net/minecraft/server/commands/CommandGamemodeDefault.java b/src/main/java/net/minecraft/server/commands/CommandGamemodeDefault.java
+index 5790699b3898910dfde46f588996dd8af42757d1..2a731594c22bb266d3dfef4d64d55b9772565cca 100644
+--- a/src/main/java/net/minecraft/server/commands/CommandGamemodeDefault.java
++++ b/src/main/java/net/minecraft/server/commands/CommandGamemodeDefault.java
+@@ -43,7 +43,13 @@ public class CommandGamemodeDefault {
+                 EntityPlayer entityplayer = (EntityPlayer) iterator.next();
+ 
+                 if (entityplayer.playerInteractManager.getGameMode() != enumgamemode) {
+-                    entityplayer.a(enumgamemode);
++                    // Paper start - handle event cancelling the change
++                    org.bukkit.event.player.PlayerGameModeChangeEvent event = entityplayer.setGamemode(enumgamemode, org.bukkit.event.player.PlayerGameModeChangeEvent.Cause.DEFAULT_GAMEMODE, net.kyori.adventure.text.Component.text("Failed to set the gamemode of '" + entityplayer.getName() + "'", net.kyori.adventure.text.format.NamedTextColor.RED));
++                    if (event != null && event.isCancelled()) {
++                        commandlistenerwrapper.sendMessage(io.papermc.paper.adventure.PaperAdventure.asVanilla(event.cancelMessage()), false);
++                        continue;
++                    }
++                    // Paper end
+                     ++i;
+                 }
+             }
+diff --git a/src/main/java/net/minecraft/server/level/EntityPlayer.java b/src/main/java/net/minecraft/server/level/EntityPlayer.java
+index 8683d6ddc5d8ce4a302fa6e2665b7dcd64f6a00a..ce4493f11e82d20cc0ec20fd01a356aaeccdefbc 100644
+--- a/src/main/java/net/minecraft/server/level/EntityPlayer.java
++++ b/src/main/java/net/minecraft/server/level/EntityPlayer.java
+@@ -396,7 +396,14 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
+         if (this.locY() > 300) this.setPositionRaw(locX(), 257, locZ()); // Paper - bring down to a saner Y level if out of world
+         if (nbttagcompound.hasKeyOfType("playerGameType", 99)) {
+             if (this.getMinecraftServer().getForceGamemode()) {
++                // Paper start - call PlayerGameModeChangeEvent on join for players that do not have the correct gamemode
++                if (this.getMinecraftServer().getGamemode() != EnumGamemode.getById(nbttagcompound.getInt("playerGameType"))) {
```

Comment:
Pretty sure you shouldn't completely skip calling the method since it initially sets the gamemode state on the server and sends it to the client

Reply:
yep, caught one of those but not the other. Added an else there to set the initial state.